### ISSUE TITLE
Explore alternatives to subclassing SoftwareComponentInternal in the Gradle plugin

### ISSUE THREAD

**Issue by wilkinsona** [spring-projects/spring-boot#9153]
Seeing what @johnrengelman has [recently done in the Shadow plugin](https://github.com/johnrengelman/shadow/commit/f0e3eedd305050822a31b8bfc834fa9d56414a56) makes me think we could do something that doesn't depend on internal API too. The Shadow plugin was extending `JavaLibrary` but the problem was essentially the same and I think John's solution may be applicable to Boot's plugin too.

**Comment by wilkinsona** [comment]
The approach taken in the Shadow plugin means that the component is no longer added to the project's `SoftwareComponentsContainer`. This hurts its discoverability so I'm not keen on doing that. 

The internal API that we're using is `SoftwareComponentInternal` a sub-interface of the `SoftwareComponent` interface that is public API. We can't implement the latter as `DefaultMavenPublication` expects every `SoftwareComponent` implementation to actually be a `SoftwareComponentInternal` implementation.

In short, I think we're going to have to live with the current arrangement for now.

**Comment by johnrengelman** [comment]
There is an open issue against Gradle for this as well - https://github.com/gradle/gradle/issues/1842
The problem here is that `SoftwareComponent` is public but is automatically converted to `SoftwareComponentInternal` inside the publishing code, so trying to use that public API actually doesn't work.

Implementing `SoftwareComponentInternal` is going to lock use of the plugin to Gradle 3.4+ because of the introduction of the internal `UsageContext` api.

**Comment by wilkinsona** [comment]
Thanks for the link to the Gradle issue, @johnrengelman

> Implementing SoftwareComponentInternal is going to lock use of the plugin to Gradle 3.4+ because of the introduction of the internal UsageContext api

We've already decided to require at least Gradle 3.4 so we're OK here (for now at least).

**Comment by wilkinsona** [comment]
In light of the breaking change in 4.0 and the news that gradle/gradle#1842 won't be addressed in Gradle 4.0, I think we should reconsider this. Given that a Spring Boot fat jar or war typically doesn't have any dependencies, configuring a `MavenPublication` by passing in the `BootJar` or `BootWar` task to its `artifact` method is as straightforward as using `from` with a custom software component.
